distributed/worker.py

```diff
@@ -1535,43 +1535,43 @@ async def close(  # type: ignore
             logger.info("Closed worker has not yet started: %s", self.status)
         if not executor_wait:
             logger.info("Not waiting on executor to close")
-        self.status = Status.closing
 
-        # Stop callbacks before giving up control in any `await`.
-        # We don't want to heartbeat while closing.
-        for pc in self.periodic_callbacks.values():
-            pc.stop()
+        # This also informs the scheduler about the status update
+        self.status = Status.closing
+        setproctitle("dask worker [closing]")
 
-        self.stop()
+        if nanny and self.nanny:
+            with self.rpc(self.nanny) as r:
+                await r.close_gracefully(reason=reason)
 
         # Cancel async instructions
         await BaseWorker.close(self, timeout=timeout)
 
-        for preload in self.preloads:
-            try:
-                await preload.teardown()
-            except Exception:
-                logger.exception("Failed to tear down preload")
+        teardowns = [
+            plugin.teardown(self)
+            for plugin in self.plugins.values()
+            if hasattr(plugin, "teardown")
+        ]
+        await asyncio.gather(*(td for td in teardowns if isawaitable(td)))
 
         for extension in self.extensions.values():
             if hasattr(extension, "close"):
                 result = extension.close()
                 if isawaitable(result):
                     await result
 
-        if nanny and self.nanny:
-            with self.rpc(self.nanny) as r:
-                await r.close_gracefully(reason=reason)
+        self.stop_services()
 
-        setproctitle("dask worker [closing]")
+        for preload in self.preloads:
+            try:
+                await preload.teardown()
+            except Exception:
+                logger.exception("Failed to tear down preload")
```

Comment:
Preloads are an interesting case. They are processed right after `ServerNode` is _initialized_ (i.e. not started). Therefore, during startup, preloads don't have access to comms and there could be a case for delaying this even further.

I don't think this is super relevant, though

Reply:
Agreed, this placement feels a bit shaky. I'd leave it here _for now_ until it becomes relevant.